## Dimention Reduction and PCA

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import math
from matplotlib.mlab import PCA as mlabPCA
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sqlalchemy import create_engine
import warnings

warnings.filterwarnings('ignore')

In [80]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'useducation'
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

us_education = pd.read_sql_query('select * from useducation',con=engine)

engine.dispose()

In [81]:
us_education.head()

PRIMARY_KEY       STATE  YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0     1992_ALABAMA     ALABAMA  1992     NaN      2678885.0         304177.0   
1      1992_ALASKA      ALASKA  1992     NaN      1049591.0         106780.0   
2     1992_ARIZONA     ARIZONA  1992     NaN      3258079.0         297888.0   
3    1992_ARKANSAS    ARKANSAS  1992     NaN      1711959.0         178571.0   
4  1992_CALIFORNIA  CALIFORNIA  1992     NaN     26260025.0        2072470.0   

   STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0      1659028.0       715680.0          2653798.0                1481703.0   
1       720711.0       222100.0           972488.0                 498362.0   
2      1369815.0      1590376.0          3401580.0                1435908.0   
3       958785.0       574603.0          1743022.0                 964323.0   
4     16546514.0      7641041.0         27138832.0               14358922.0   

   ...  GRADES_4_G  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  GRADES_9_12_G  \
0  ...     57948.0     58025.0      41167.0      471564.0       196386.0   
1  ...      9748.0      8789.0       6714.0       79117.0        30847.0   
2  ...     55433.0     49081.0      37410.0      437127.0       175210.0   
3  ...     34632.0     36011.0      27651.0      281338.0       123113.0   
4  ...    418418.0    363296.0     270675.0     3286034.0      1372011.0   

   GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
0      676174.0        208.327876        252.187522           207.963517   
1      112335.0               NaN               NaN                  NaN   
2      614881.0        215.253932        265.366278           206.212716   
3      405259.0        210.206028        256.312090           208.634458   
4     4717112.0        208.398961        260.892247           196.764414   

   AVG_READING_8_SCORE  
0                  NaN  
1           258.859712  
2           262.169895  
3           264.619665  
4                  NaN  

[5 rows x 25 columns]

In [82]:
us_education.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1492 entries, 0 to 1491
Data columns (total 25 columns):
PRIMARY_KEY                     1492 non-null object
STATE                           1492 non-null object
YEAR                            1492 non-null int64
ENROLL                          1229 non-null float64
TOTAL_REVENUE                   1280 non-null float64
FEDERAL_REVENUE                 1280 non-null float64
STATE_REVENUE                   1280 non-null float64
LOCAL_REVENUE                   1280 non-null float64
TOTAL_EXPENDITURE               1280 non-null float64
INSTRUCTION_EXPENDITURE         1280 non-null float64
SUPPORT_SERVICES_EXPENDITURE    1280 non-null float64
OTHER_EXPENDITURE               1229 non-null float64
CAPITAL_OUTLAY_EXPENDITURE      1280 non-null float64
GRADES_PK_G                     1319 non-null float64
GRADES_KG_G                     1360 non-null float64
GRADES_4_G                      1361 non-null float64
GRADES_8_G                      1

First lets fill the missing values for all the columns except for the AVG_* columns.To accomplish that the best way is using interpolation beacause it is a time series data with respect to the states.

In [83]:
columns_to_fill = ['PRIMARY_KEY', 'STATE', 'YEAR', 'ENROLL', 'TOTAL_REVENUE', 'FEDERAL_REVENUE',
             'STATE_REVENUE', 'LOCAL_REVENUE', 'TOTAL_EXPENDITURE',
             'INSTRUCTION_EXPENDITURE', 'SUPPORT_SERVICES_EXPENDITURE',
             'OTHER_EXPENDITURE', 'CAPITAL_OUTLAY_EXPENDITURE', 'GRADES_PK_G',
             'GRADES_KG_G', 'GRADES_4_G', 'GRADES_8_G', 'GRADES_12_G', 'GRADES_1_8_G',
             'GRADES_9_12_G', 'GRADES_ALL_G']
for state in us_education['STATE'].unique():
    us_education.loc[us_education['STATE'] == state
                     , columns_to_fill] = us_education.loc[us_education['STATE'] == state
                                                           , columns_to_fill].interpolate()
us_education.dropna(inplace = True)    

Then we drop the rest of the null values;

Now that our data is clean and ready for next step in the EDA process;lets answer the questions for this assignment.

**1. Create a new score variable from the weighted averages of all score variables in the datasets. Notice that the number of students in the 4th grade isn't the same as the number of students in the 8th grade. So, you should appropriately weigh the scores!.**

In [84]:
us_education.head()

PRIMARY_KEY       STATE  YEAR     ENROLL  TOTAL_REVENUE  \
209     1996_ALABAMA     ALABAMA  1996   735912.0      3365113.0   
211     1996_ARIZONA     ARIZONA  1996   764681.0      4143304.0   
212    1996_ARKANSAS    ARKANSAS  1996   452907.0      2183384.0   
213  1996_CALIFORNIA  CALIFORNIA  1996  5460484.0     31282981.0   
214    1996_COLORADO    COLORADO  1996   655679.0      3891203.0   

     FEDERAL_REVENUE  STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
209         334858.0      2138717.0       891538.0          3572283.0   
211         378917.0      1798363.0      1966024.0          4391555.0   
212         192152.0      1296247.0       694985.0          2297381.0   
213        2603882.0     17061474.0     11617625.0         31561692.0   
214         194998.0      1664676.0      2031529.0          4075165.0   

     INSTRUCTION_EXPENDITURE  ...  GRADES_4_G  GRADES_8_G  GRADES_12_G  \
209                1987018.0  ...     57098.0     58305.0      42510.0   
211                1860007.0  ...     62793.0     60514.0      42041.0   
212                1264275.0  ...     33855.0     37064.0      27613.0   
213               16880730.0  ...    435217.0    406456.0     298669.0   
214                2052264.0  ...     52524.0     52269.0      37179.0   

     GRADES_1_8_G  GRADES_9_12_G  GRADES_ALL_G  AVG_MATH_4_SCORE  \
209      469370.0       207980.0      686983.0        211.646974   
211      511925.0       210289.0      726709.0        217.575940   
212      283883.0       131997.0      417860.0        215.846436   
213     3516572.0      1525618.0     5115423.0        209.129083   
214      423510.0       185700.0      621730.0        225.805858   

     AVG_MATH_8_SCORE  AVG_READING_4_SCORE  AVG_READING_8_SCORE  
209        256.594863           210.920767           260.276441  
211        267.874834           206.529788           263.283638  
212        261.652745           208.837716           271.834590  
213        262.772478           201.912591           269.197451  
214        275.608214           221.653243           268.168154  

[5 rows x 25 columns]

In [92]:
us_education['weighed_scores_avg'] = (us_education['GRADES_4_G'] 
                                      * ((us_education['AVG_MATH_4_SCORE'] + us_education['AVG_READING_4_SCORE']) 
                                      * 0.5) + us_education['GRADES_8_G'] 
                                      * ((us_education['AVG_MATH_8_SCORE'] + us_education['AVG_READING_8_SCORE']) 
                                      * 0.5))/(us_education['GRADES_4_G'] + us_education['GRADES_8_G'])

In [93]:
us_education.head()

PRIMARY_KEY       STATE  YEAR     ENROLL  TOTAL_REVENUE  \
209     1996_ALABAMA     ALABAMA  1996   735912.0      3365113.0   
211     1996_ARIZONA     ARIZONA  1996   764681.0      4143304.0   
212    1996_ARKANSAS    ARKANSAS  1996   452907.0      2183384.0   
213  1996_CALIFORNIA  CALIFORNIA  1996  5460484.0     31282981.0   
214    1996_COLORADO    COLORADO  1996   655679.0      3891203.0   

     FEDERAL_REVENUE  STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
209         334858.0      2138717.0       891538.0          3572283.0   
211         378917.0      1798363.0      1966024.0          4391555.0   
212         192152.0      1296247.0       694985.0          2297381.0   
213        2603882.0     17061474.0     11617625.0         31561692.0   
214         194998.0      1664676.0      2031529.0          4075165.0   

     INSTRUCTION_EXPENDITURE  ...  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  \
209                1987018.0  ...     58305.0      42510.0      469370.0   
211                1860007.0  ...     60514.0      42041.0      511925.0   
212                1264275.0  ...     37064.0      27613.0      283883.0   
213               16880730.0  ...    406456.0     298669.0     3516572.0   
214                2052264.0  ...     52269.0      37179.0      423510.0   

     GRADES_9_12_G  GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  \
209       207980.0      686983.0        211.646974        256.594863   
211       210289.0      726709.0        217.575940        267.874834   
212       131997.0      417860.0        215.846436        261.652745   
213      1525618.0     5115423.0        209.129083        262.772478   
214       185700.0      621730.0        225.805858        275.608214   

     AVG_READING_4_SCORE  AVG_READING_8_SCORE  weighed_scores_avg  
209           210.920767           260.276441          235.106342  
211           206.529788           263.283638          238.321404  
212           208.837716           271.834590          240.773675  
213           201.912591           269.197451          234.719834  
214           221.653243           268.168154          247.750273  

[5 rows x 26 columns]

**2. What are the correlations between this newly created score variable and the expenditure types? Which 1 of the expenditure types is more correlated than the others?**

In [94]:
us_education.columns

Index(['PRIMARY_KEY', 'STATE', 'YEAR', 'ENROLL', 'TOTAL_REVENUE',
       'FEDERAL_REVENUE', 'STATE_REVENUE', 'LOCAL_REVENUE',
       'TOTAL_EXPENDITURE', 'INSTRUCTION_EXPENDITURE',
       'SUPPORT_SERVICES_EXPENDITURE', 'OTHER_EXPENDITURE',
       'CAPITAL_OUTLAY_EXPENDITURE', 'GRADES_PK_G', 'GRADES_KG_G',
       'GRADES_4_G', 'GRADES_8_G', 'GRADES_12_G', 'GRADES_1_8_G',
       'GRADES_9_12_G', 'GRADES_ALL_G', 'AVG_MATH_4_SCORE', 'AVG_MATH_8_SCORE',
       'AVG_READING_4_SCORE', 'AVG_READING_8_SCORE', 'weighed_scores_avg'],
      dtype='object')

In [95]:
us_education[['weighed_scores_avg', 'TOTAL_EXPENDITURE','INSTRUCTION_EXPENDITURE', 'SUPPORT_SERVICES_EXPENDITURE'
              , 'OTHER_EXPENDITURE', 'CAPITAL_OUTLAY_EXPENDITURE']].corr()

weighed_scores_avg  TOTAL_EXPENDITURE  \
weighed_scores_avg                      1.000000           0.074155   
TOTAL_EXPENDITURE                       0.074155           1.000000   
INSTRUCTION_EXPENDITURE                 0.087386           0.990255   
SUPPORT_SERVICES_EXPENDITURE            0.079002           0.992988   
OTHER_EXPENDITURE                       0.010959           0.943702   
CAPITAL_OUTLAY_EXPENDITURE             -0.005389           0.925784   

                              INSTRUCTION_EXPENDITURE  \
weighed_scores_avg                           0.087386   
TOTAL_EXPENDITURE                            0.990255   
INSTRUCTION_EXPENDITURE                      1.000000   
SUPPORT_SERVICES_EXPENDITURE                 0.974399   
OTHER_EXPENDITURE                            0.902173   
CAPITAL_OUTLAY_EXPENDITURE                   0.884798   

                              SUPPORT_SERVICES_EXPENDITURE  OTHER_EXPENDITURE  \
weighed_scores_avg                                0.079002           0.010959   
TOTAL_EXPENDITURE                                 0.992988           0.943702   
INSTRUCTION_EXPENDITURE                           0.974399           0.902173   
SUPPORT_SERVICES_EXPENDITURE                      1.000000           0.952282   
OTHER_EXPENDITURE                                 0.952282           1.000000   
CAPITAL_OUTLAY_EXPENDITURE                        0.911245           0.918188   

                              CAPITAL_OUTLAY_EXPENDITURE  
weighed_scores_avg                             -0.005389  
TOTAL_EXPENDITURE                               0.925784  
INSTRUCTION_EXPENDITURE                         0.884798  
SUPPORT_SERVICES_EXPENDITURE                    0.911245  
OTHER_EXPENDITURE                               0.918188  
CAPITAL_OUTLAY_EXPENDITURE                      1.000000

from the correlation table in the above we can see that 'INSTRUCTION_EXPENDITURE' is more related to the 'weiged_scores_avg' with a correlation value of 0.087386.

**3. Now, apply PCA to the 4 expenditure types. How much of the total variance is explained by the 1st component?**

The four expenditure types are "INSTRUCTION_EXPENDITURE", "SUPPORT_SERVICES_EXPENDITURE","OTHER_EXPENDITURE" and "CAPITAL_OUTLAY_EXPENDITURE" with out the total expenditure. Therefore lets create our X.

In [97]:
X = us_education[["INSTRUCTION_EXPENDITURE", "SUPPORT_SERVICES_EXPENDITURE",
                  "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE"]]

Then we standardize our X(transform it to standard normal)

In [99]:
X = StandardScaler().fit_transform(X)
#lets select number of principal components which in our case 1.
sklearn_pca = PCA(n_components=1)
#create the first principal component 'pca_1' as new column in our main data set.
us_education["pca_1"] = sklearn_pca.fit_transform(X)

In [105]:
print('{}, ratio which is {}% of the total variance in the data was explained by the first component.'
      .format(sklearn_pca.explained_variance_ratio_[0], sklearn_pca.explained_variance_ratio_[0] * 100))

0.9430174977604754, ratio which is 94.30174977604754% of the total variance in the data was explained by the first component.


**4. What is the correlation between the overall score variable and the 1st principal component?**

In [106]:
us_education[['weighed_scores_avg', 'pca_1']].corr()

weighed_scores_avg     pca_1
weighed_scores_avg            1.000000  0.044596
pca_1                         0.044596  1.000000

it is 0.044596.

**5. If you were to choose the best variables for your model, would you prefer using the 1st principal component instead of the expenditure variables? Why?**

From our previous correlation table of expenditures versus the weighed_scores_avg variable(outcome), 'INSTRUCTION_EXPENDITURE' variable has the highest correlation of 0.087386 which is still higher than our 1st principal component.Therefore 'INSTRUCTION_EXPENDITURE' is still the best variable for our model.